feat(gatsby): add opt out flag SKIP_PLUGIN_OPTION_VALIDATION to bypass validation

gillkyle opened the pull request:
<!--
  Have any questions? Check out the contributing docs at https://gatsby.dev/contribute, or
  ask in this Pull Request and a Gatsby maintainer will be happy to help :)
-->

<!--
  Is this a blog post? Check out the docs at https://www.gatsbyjs.com/contributing/blog-contributions/, and please mention if the blog post is pre-approved
  by someone from Gatsby.
-->

## Description

<!-- Write a brief description of the changes introduced by this PR -->

Some users of the latest versions of Gatsby have run into problems with undocumented plugin options we didn't catch when adding schemas. When schema validation fails, the `build` and `develop` process errors out (`process.exit(1)`). To give an unblocking workaround for users frustrated by this behavior, this change allows the `SKIP_PLUGIN_OPTION_VALIDATION` to be set, a la:

```
SKIP_PLUGIN_OPTION_VALIDATION=true gatsby develop
```

To skip the validation and instead issue a warning that looks like this:

![image](https://user-images.githubusercontent.com/21114044/98416082-afd40880-203b-11eb-97ca-34924b2fd249.png)

An alternative approach to implementing this would be warning that options not included in the schema have separate warnings issued for them, rather than erroring. It looks like [this should be possible with Joi](https://joi.dev/api/?v=17.3.0#anyvalidateasyncvalue-options) (see the `warnings` option) but I couldn't get it to work and figured this would be a relatively quick way to get around this without digging into Joi and have a similar effect.


### Documentation

<!--
  Where is this feature or API documented?

  - If docs exist:
    - Update any references, if relevant. This includes Guides and Gatsby Internals docs.
  - If no docs exist:
    - Create a stub for documentation including bullet points for how to use the feature, code snippets (including from happy path tests), etc.
  - Tag @gatsbyjs/documentation for review, pairing, polishing of the documentation
-->

I added a note to the [plugin options guide](https://www.gatsbyjs.com/docs/configuring-usage-with-plugin-options/#how-to-validate-plugin-options)

## Related Issues

<!--
  Link to the issue that is fixed by this PR (if there is one)
  e.g. Fixes #1234

  Link to an issue that is partially addressed by this PR (if there are any)
  e.g. Addresses #1234

  Link to related issues (if there are any)
  e.g. Related to #1234
-->

Related to https://github.com/gatsbyjs/gatsby/issues/27839

KyleAMathews commented:
How about when we error we also tell people about this flag so they can quickly get around bugs while it's getting fixed?

KyleAMathews commented:
Also encourage them to file bugs if they believe the validation is in error.

mxstbr commented on packages/gatsby/src/bootstrap/load-plugins/validate.ts:
```diff
@@ -271,6 +271,13 @@ export async function validateConfigPluginsOptions(
+  if (process.env.SKIP_PLUGIN_OPTION_VALIDATION) {
+    reporter.warn(`you have enabled the SKIP_PLUGIN_OPTION_VALIDATION flag,`)
+    reporter.warn(
+      `meaning Gatsby will not enforce types based on plugin's option schemas.`
```
```suggestion
      `so Gatsby will not verify your plugin configurations.`
```

mxstbr reviewed (commented):
I like what @kylemathews suggested above!

ascorbic commented on docs/docs/configuring-usage-with-plugin-options.md:
````diff
@@ -276,6 +276,14 @@ describe(`pluginOptionsSchema`, () => {
+```shell
+SKIP_PLUGIN_OPTION_VALIDATION=true gatsby develop
+```
+
````
```suggestion
Please also [submit an issue](https://github.com/gatsbyjs/gatsby/issues) if you believe the options are failing incorrectly.
```

wardpeet commented:
This is dangerous. By bypassing validation, default values will not be set and cause more troubles. We should allow unknown values instead.

mxstbr commented:
> We should allow unknown values instead.

There could be other issues with the schemas. I think the better solution is to not `process.exit` if the flag is set (i.e. still set the default values), but still show the errors in the console?

https://github.com/gatsbyjs/gatsby/blob/557139ed20fb18db08686db42284511ab8f9bdfd/packages/gatsby/src/bootstrap/load-plugins/validate.ts#L282-L284

gillkyle commented:
Okay I tried to make updates based on everyone's feedback and I _think_ this would address them for now.

Here's what happens running with and without the flag in a `hello-world` site with one erroneous plugin installed:

![CleanShot 2020-11-09 at 09 07 59@2x](https://user-images.githubusercontent.com/21114044/98566542-1a6b8b00-226c-11eb-9814-5899be368e33.png)

Things I did:

- [x] switched `process.exit` to `reporter.panic` (a suggestion from Matt)
- [x] added the suggestion left in code review, encouraging users to file issues and tweaking error wording
- [x] moved the check of the flag that bypasses logic **after** the validation runs (based off of @wardpeet's comment) so that validation still runs
- [x] added another `if` around the reporter warning so errors are not logged if you have the flag set (it will only show the warnings and not the specific error message for each erroneous plugin)

Am I still missing things?

gillkyle commented:
I'm going to look into the failing tests now, it looks like this might alter the assertions that some of them make..

mxstbr commented on docs/docs/configuring-usage-with-plugin-options.md:
````diff
@@ -276,6 +276,16 @@ describe(`pluginOptionsSchema`, () => {
+If you are running your site and run into an error trying to develop or build your site due to plugin options failing that are undocumented or seem erroneous, you can add the `SKIP_PLUGIN_OPTION_VALIDATION` flag set to `true`. It can be prepended to your develop and build scripts, included as an [environment variable](/docs/environment-variables/) in a .env file, or included before your `gatsby d ...
+
+```shell
+SKIP_PLUGIN_OPTION_VALIDATION=true gatsby develop
````
**Bikeshedding, feel free to ignore/postpone**

I'd rather not add this to the docs so explicitly because we don't _really_ want people to use this _unless_ they have a blocking error.

I would prefer if we add that information to the error log if they have an error with an ominous warning, for example something like:

```
success open and validate gatsby-configs - 0.017s

 ERROR #11331  PLUGIN

Invalid plugin options for "gatsby-plugin-manifest":

- "name" must be a string

If you are certain your configuration is correct, you can disable configuration validation with the SKIP_PLUGIN_OPTION_VALIDATION environment variable. Be careful, this could have unintended side effects!

not finished load plugins - 2.391s
```

laurieontech replied:
Ooh, this is a fun one. Ya, I totally see what you mean. At the same time, how often do people skip error messages and go googling? No strong feelings, just a weird thought exercise.

mxstbr commented on packages/gatsby/src/bootstrap/load-plugins/validate.ts:
```diff
@@ -280,7 +282,13 @@ export async function validateConfigPluginsOptions(
   if (errors > 0) {
-    process.exit(1)
+    if (process.env.SKIP_PLUGIN_OPTION_VALIDATION) {
+      reporter.warn(`you have enabled the SKIP_PLUGIN_OPTION_VALIDATION flag,`)
```
Nitpick: I'd rather we mention explicitly everytime that there is incorrect configuration, something like:

```suggestion
      reporter.warn(`your gatsby-config.js contains invalid plugin options`)
      reporter.warn(`because you have enabled the SKIP_PLUGIN_OPTION_VALIDATION flag, develop will try continuing`)
      reporter.warn(`but this could have unintended side effects and break your site`)
```

mxstbr reviewed (approved):
Looks fantastic to me! Love the behaviour as well :100: I've added some bikeshedding nitpicks inline, feel free to postpone or implement them now — up to you.

gillkyle commented:
With https://github.com/gatsbyjs/gatsby/pull/27938 merged, which in my opinion is the ideal solution to this problem, I'm going to pause on this for now. I suspect we may not need to add this flag at all since warnings will surface when a plugin has an unknown option which is the majority of problems that have been encountered so far. 

Max noted it's possible there could be some errors not caught by warnings but I'm not sure we've seen too many of those to justify needing the flag if that PR covers most issues. I'm all ears if someone else feels more strongly but I'm going to wait a little before closing this, just to see if anyone has more justification for it.

LekoArts commented:
@gillkyle Can this be closed then?

gillkyle commented:
Thanks for the reminder @LekoArts! It can indeed, Max shipped the warnings PR so this is irrelevant 🙂
